src/charm.py

```diff
@@ -5,80 +5,160 @@
+
+    certificate: str
+    private_key: str
+
+    def to_dict(self) -> dict[str, str]:
+        """Return a dict version of the secret."""
+        return {"certificate": self.certificate, "private-key": self.private_key}
+
 
 class GocertCharm(ops.CharmBase):
     """Charmed Gocert."""
 
     def __init__(self, framework: ops.Framework):
         super().__init__(framework)
+
         self.container = self.unit.get_container("gocert")
         self.tls = TLSCertificatesProvidesV3(self, relationship_name="certificates")
 
         self.port = 2111
 
-        framework.observe(self.on["gocert"].pebble_ready, self._install)
-        framework.observe(self.on["gocert"].pebble_custom_notice, self._on_gocert_notify)
-        framework.observe(self.tls.on.certificate_creation_request, self._on_new_certificate)
+        [
+            framework.observe(event, self.configure)
+            for event in [
+                self.on["gocert"].pebble_ready,
+                self.on["gocert"].pebble_custom_notice,
+                self.tls.on.certificate_creation_request,
+                self.on.config_storage_attached,
+                self.on.database_storage_attached,
+                self.on.config_changed,
+                self.on.update_status,
+                self.on.start,
+            ]
+        ]
+        framework.observe(self.on.collect_app_status, self._on_collect_status)
         framework.observe(self.on.collect_unit_status, self._on_collect_status)
 
-    def _install(self, event: ops.PebbleReadyEvent):
+    def configure(self, event: ops.EventBase):
+        """Handle configuration events."""
+        if not self.unit.is_leader():
+            return
+        if not self.container.can_connect():
+            return
+        self._configure_gocert_config_file()
+        self._configure_access_certificates()
+        self._configure_certificate_relations()
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent):
+        if not self.unit.is_leader():
+            event.add_status(ops.WaitingStatus("multiple units not supported"))
+            return
+        if not self.container.can_connect():
+            event.add_status(ops.WaitingStatus("container not yet connectable"))
+            return
+        if not self._storages_attached():
+            event.add_status(ops.WaitingStatus("storages not yet available"))
+            return
+        if not self._self_signed_certificates_generated():
+            event.add_status(ops.WaitingStatus("certificates not yet created"))
+            return
+        if not self._gocert_available():
+            event.add_status(ops.WaitingStatus("GoCert server not yet available"))
+            return
+        if not self._gocert_initialized():
+            event.add_status(ops.BlockedStatus("Please initialize GoCert"))
+            return
+        event.add_status(ops.ActiveStatus())
+
+    def _on_gocert_notify(self, event: ops.PebbleCustomNoticeEvent):
```

Comment:
This likely deserves a docstring

Reply:
oops, the contents of this function should be in the configure_certificate_relations function, which has the docstring. this function is deleted.